### PR TITLE
Add aten::nll_loss_forward op lowering.

### PR DESCRIPTION
**Pull request by pashu123**
The op lowering has been added as a part of `torch-lower-to-linalg`
pass. This takes care of ignore_index but the weight and reduction
operand is still to be accounted for.

Signed-off-by: Prashant Kumar <prashant@nod-labs.com>

### CODE REVIEW

**Comment by pashu123** [comment]
@cathyzhyi Apart from `nll_loss_forward` op there is something related to addCmul and addCdiv that is addressed. The errors popped due to name change. Could you take a look at them?

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 1]
```diff
@@ -923,6 +930,38 @@ ChangeResult TypeAnalyzer::visitAtenSqueezeOp(
...
+      ValueKnowledge::getNotNonePessimisticValueState(op.getContext());
+  knowledge2.dtype = self.dtype;
+  knowledge2.sizes.resize(0, kUnknownSize);
+  knowledge2.hasSizes = true;
+
+  if (self.hasSizes &&
+      matchPattern(op.reduction(), m_TorchConstantInt(&reduction))) {
+    resultRank = reduction == 1 ? resultRank - 1 : resultRank;
+  }
+  knowledge.sizes.resize(resultRank-1, kUnknownSize);
```
Comment (cathyzhyi) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:958]: I think the `outputKnowledge` can be more accurate. Use `N*C*d0*d1` as an example. For none reduction case, it's the C dimension reduced. For reduction case, it's the MiniBatch(N) dimension and C dimension both got reduced.

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

**Comment by cathyzhyi** [comment]
> @cathyzhyi Apart from `nll_loss_forward` op there is something related to addCmul and addCdiv that is addressed. The errors popped due to name change. Could you take a look at them?

Are the changes from picking up newer pytorch? Are you using the pytorch nightly build?

**Comment by cathyzhyi** [comment]
> > @cathyzhyi Apart from `nll_loss_forward` op there is something related to addCmul and addCdiv that is addressed. The errors popped due to name change. Could you take a look at them?
> 
> Are the changes from picking up newer pytorch? Are you using the pytorch nightly build?

Ha, I see those as well. Thanks for fixing those.

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 2]
```diff
@@ -1167,6 +1167,91 @@ class ConvertAtenDropoutOp : public OpConversionPattern<AtenDropoutOp> {
...
+    if (!failed(checkNotNone(rewriter, op, weight)))
+      return failure();
+
+    Value ignoreIndex = adaptor.ignore_index();
+    Value ignoreIndexVal = castIntToIndex(rewriter, loc, ignoreIndex);
+
+    unsigned inputRank = input.getType().cast<RankedTensorType>().getRank();
+    unsigned targetRank = target.getType().cast<RankedTensorType>().getRank();
+
+    // TODO: Cases with targetRank != 1 where `Mean` reduction is required.
```
[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1200]: https://github.com/pytorch/pytorch/blob/e0fb228e0346a01f16592a1ced106e0f7d5efafd/aten/src/ATen/native/LossNLL.cpp#L26

[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1200]: The kernel checks that the target rank <= 1.

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp (hunk 1 above)
Comment (pashu123) [reply lib/Dialect/Torch/Transforms/RefineTypes.cpp:958]: https://github.com/pytorch/pytorch/blob/e0fb228e0346a01f16592a1ced106e0f7d5efafd/aten/src/ATen/native/LossNLL.cpp#L24  The self dim is either 1 or 2.

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:1210]: Done.

File: e2e_testing/torchscript/nll_loss.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    super().__init__()
+
+  @export
+  @annotate_args([
+      None,
+      ([-1, -1], torch.float32, True),
+      ([-1], torch.int64, True),
+  ])
+  # Here the 2nd index is ignored.
+  def forward(self, x, y):
```
[the remark this answers is not in the document]
Comment (pashu123) [reply e2e_testing/torchscript/nll_loss.py:27]: Done.

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 4]
```diff
@@ -927,6 +934,38 @@ ChangeResult TypeAnalyzer::visitAtenSqueezeOp(
   return getLatticeElement(op.getResult()).join(knowledge);
 }
 
+ChangeResult TypeAnalyzer::visitAtenNllLossForwardOp(
+    AtenNllLossForwardOp op,
+    ArrayRef<LatticeElement<ValueKnowledge> *> operands) {
+  auto self = operands[0]->getValue();
+  auto target = operands[1]->getValue();
+  auto weight = operands[2]->getValue();
```
Comment (gprateek93) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:942]: These variables are unused?

File: e2e_testing/torchscript/nll_loss.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  def __init__(self):
+    super().__init__()
+
+  @export
+  @annotate_args([
+      None,
+      ([-1, -1], torch.float32, True),
+      ([-1], torch.int64, True),
+  ])
+  # Here the 2nd index is ignored.
```
Comment (silvasean) [inline e2e_testing/torchscript/nll_loss.py:26]: nit: instead of a comment, put this in the module/test name, e.g. NllLossModule_ignore_index

File: e2e_testing/torchscript/nll_loss.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                                           reduction=0,
+                                           ignore_index=2)[0]
+
+
+@register_test_case(module_factory=lambda: NllLossModule())
+def NllLossModule_basic(module, tu: TestUtils):
+  module.forward(tu.rand(2, 3), torch.tensor([0, 1]))
+
+
+class NllLossModule1(torch.nn.Module):
```
Comment (silvasean) [inline e2e_testing/torchscript/nll_loss.py:40]: NllLossModule_ignore_index_out_of_bounds

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 7]
```diff
@@ -927,6 +934,36 @@ ChangeResult TypeAnalyzer::visitAtenSqueezeOp(
...
+  auto self = operands[0]->getValue();
+  auto outputKnowledge =
+      ValueKnowledge::getNotNonePessimisticValueState(op.getContext());
+
+  // Contains Knowledge of shape and dtype for the 1st result.
+  outputKnowledge.dtype = self.dtype;
+  int64_t reduction;
+  unsigned resultRank = self.sizes.size();
+
+  // Contains Knowledge of shape and dtype for the 2st result.
```
Comment (silvasean) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:949]: nit: 2nd

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 8]
```diff
@@ -927,6 +934,36 @@ ChangeResult TypeAnalyzer::visitAtenSqueezeOp(
...
+      ValueKnowledge::getNotNonePessimisticValueState(op.getContext());
+  totalWeightKnowledge.dtype = self.dtype;
+  totalWeightKnowledge.sizes.resize(0, kUnknownSize);
+  totalWeightKnowledge.hasSizes = true;
+
+  if (self.hasSizes &&
+      matchPattern(op.reduction(), m_TorchConstantInt(&reduction))) {
+    resultRank = reduction == 1 ? resultRank - 1 : resultRank;
+  }
+  outputKnowledge.sizes.resize(resultRank-1, kUnknownSize);
```
Comment (silvasean) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:960]: nit: clang-format `resultRank - 1`.

File: lib/Dialect/Torch/Transforms/RefineTypes.cpp [hunk 9]
```diff
@@ -927,6 +934,36 @@ ChangeResult TypeAnalyzer::visitAtenSqueezeOp(
...
+  // Contains Knowledge of shape and dtype for the 2st result.
+  auto totalWeightKnowledge =
+      ValueKnowledge::getNotNonePessimisticValueState(op.getContext());
+  totalWeightKnowledge.dtype = self.dtype;
+  totalWeightKnowledge.sizes.resize(0, kUnknownSize);
+  totalWeightKnowledge.hasSizes = true;
+
+  if (self.hasSizes &&
+      matchPattern(op.reduction(), m_TorchConstantInt(&reduction))) {
+    resultRank = reduction == 1 ? resultRank - 1 : resultRank;
```
Comment (silvasean) [inline lib/Dialect/Torch/Transforms/RefineTypes.cpp:958]: comment on what "1" means here.

**Review by silvasean — APPROVED** [review]
LGTM with nits.
